File: src/main/java/com/janrain/oauth2/AuthorizationDecisionKey.java
/*
 * Copyright 2012 Janrain, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.janrain.oauth2;

import com.janrain.backplane.server.ExternalizableCore;
import com.janrain.backplane2.server.config.Backplane2Config;
import com.janrain.commons.message.MessageException;
import com.janrain.commons.message.MessageField;
import com.janrain.crypto.ChannelUtil;

import java.util.*;

/**
 * @author Johnny Bufu
 */
public class AuthorizationDecisionKey extends ExternalizableCore {

    // - PUBLIC

    /**
     * Empty default constructor for AWS to use
     */
    public AuthorizationDecisionKey() { }

    public AuthorizationDecisionKey(String authCookie) throws MessageException {
        Map<String,String> data = new LinkedHashMap<String, String>();
        String key = ChannelUtil.randomString(AUTHORIZATION_DECISION_KEY_LENGTH);
        data.put(Field.KEY.getFieldName(), key);
        data.put(Field.AUTH_COOKIE.getFieldName(), authCookie);
        data.put(Field.EXPIRES.getFieldName(), Backplane2Config.ISO8601.get().format(new Date(System.currentTimeMillis() + AUTHORIZATION_DECISION_TIMEOUT_SECONDS * 1000)));
        super.init(key, data);
    }

    @Override
    public String getIdValue() {
        return get(Field.KEY);
    }

    @Override
    public Set<? extends MessageField> getFields() {
        return EnumSet.allOf(Field.class);
    }

    public static enum Field implements MessageField {

        KEY,
        AUTH_COOKIE,
        EXPIRES;

        @Override
        public String getFieldName() {
            return name();
        }

        @Override
        public boolean isRequired() {
            return true;
        }

        @Override
        public void validate(String value) throws MessageException {
            if (isRequired()) validateNotBlank(name(), value);
        }
    }

    // - PRIVATE

    private static final long serialVersionUID = -3343912147061904056L;

    private static final int AUTHORIZATION_DECISION_KEY_LENGTH = 30;
    private static final long AUTHORIZATION_DECISION_TIMEOUT_SECONDS = 300l;
}

File: src/main/scala/com/janrain/backplane/server2/oauth2/model/AuthorizationDecisionKey.scala
package com.janrain.backplane.server2.oauth2.model

import com.janrain.backplane.common.DateTimeUtils
import java.util.Date
import com.janrain.util.RandomUtils
import com.janrain.backplane.common.model.{MessageField, MessageFieldEnum, Message}

/**
 * @author Johnny Bufu
 */
class AuthorizationDecisionKey(data: Map[String,String]) extends Message(data, AuthorizationDecisionKeyFields.values) {

  def this(authCookie: String) = this(
    Map( AuthorizationDecisionKeyFields.AUTH_COOKIE.name -> authCookie,
         AuthorizationDecisionKeyFields.KEY.name
           -> RandomUtils.randomString(AuthorizationDecisionKey.AUTHORIZATION_DECISION_KEY_LENGTH),
         AuthorizationDecisionKeyFields.EXPIRES.name
           -> DateTimeUtils.ISO8601.get().format(new Date(System.currentTimeMillis() + AuthorizationDecisionKey.AUTHORIZATION_DECISION_TIMEOUT_SECONDS * 1000))
    ))

  def idField = AuthorizationDecisionKeyFields.KEY
}

object AuthorizationDecisionKey {
  private[model] final val AUTHORIZATION_DECISION_KEY_LENGTH: Int = 30
  private[model] final val AUTHORIZATION_DECISION_TIMEOUT_SECONDS: Long = 300L
}

object AuthorizationDecisionKeyFields extends MessageFieldEnum {

  type AuthorizationDecisionKeyField = EnumVal

  sealed trait EnumVal extends Value with MessageField {
    def required = true
  }

  val KEY = new AuthorizationDecisionKeyField { def name = "key" }

  val AUTH_COOKIE = new AuthorizationDecisionKeyField { def name = "auth_cookie" }

  val EXPIRES = new AuthorizationDecisionKeyField { def name = "expires" }
}
